storage/buffer.go

```diff
@@ -199,26 +194,97 @@ func (b *dbBuffer) ReadEncoded(ctx context.Context, start, end time.Time) [][]xi
 		}
 
 		unmerged := make([]xio.SegmentReader, 0, len(bucket.encoders))
-		for i := range bucket.encoders {
-			stream := bucket.encoders[i].encoder.Stream()
-			if stream == nil {
-				// TODO(r): log an error and emit a metric, this is pretty bad as this
-				// encoder should have values if "shouldRead" returned true
-				continue
+		b.readBucketStreams(ctx, bucket, now, current, func(stream xio.SegmentReader) {
+			unmerged = append(unmerged, stream)
+		})
+
+		results = append(results, unmerged)
+	})
+
+	return results
+}
+
+func (b *dbBuffer) FetchBlocks(ctx context.Context, starts []time.Time) []FetchBlockResult {
+	var res []FetchBlockResult
+
+	now := b.nowFn()
+	b.forEachBucketAsc(now, func(bucket *dbBufferBucket, current time.Time) {
+		found := false
+		// starts has ~24 items, linear search should be okay time-wise to
+		// avoid allocating a map here.
+		for _, start := range starts {
+			if start == bucket.start {
+				found = true
+				break
 			}
+		}
+		if !found {
+			return
+		}
+		readers := make([]xio.SegmentReader, 0, len(bucket.encoders))
+		b.readBucketStreams(ctx, bucket, now, current, func(stream xio.SegmentReader) {
+			readers = append(readers, stream)
+		})
+		res = append(res, newFetchBlockResult(bucket.start, readers, nil))
+	})
+
+	return res
+}
 
-			ctx.RegisterCloser(func() {
-				// Close the stream and return to pool when read done
-				stream.Close()
+func (b *dbBuffer) FetchBlocksMetadata(ctx context.Context, includeSizes bool) []block.DatabaseBlockMetadata {
+	var res []block.DatabaseBlockMetadata
+
+	now := b.nowFn()
+	b.forEachBucketAsc(now, func(bucket *dbBufferBucket, current time.Time) {
+		var pSize *int64
+		if includeSizes {
+			var size int64
+			b.readBucketStreams(ctx, bucket, now, current, func(stream xio.SegmentReader) {
+				segment := stream.Segment()
+				size += int64(len(segment.Head) + len(segment.Tail))
```

Comment:
Probably want to `stream.Close()` before the end of this method here, otherwise the segment reader won't get returned to the pool.